Python bindings for AMReX-MPMD

siddanib opened the pull request:
This PR is a **draft** that includes python bindings for Multiple Program Multiple Data (MPMD) feature of AMReX. This feature depends on MPI for the involved C++ executable and mpi4py for the involved python script. **The submitted python bindings are based on the latest AMReX_MPMD.H and AMReX_MPMD.cpp**.

Commit:
First changes related to MPMD feature in pyAMReX

siddanib committed:
Merge branch 'AMReX-Codes:development' into development

siddanib committed:
Merge branch 'AMReX-Codes:development' into development

Commit:
Two MultiFab tests with MPMD::Copier::send from both sides.

siddanib committed:
Merge branch 'AMReX-Codes:development' into development

siddanib committed:
Merge branch 'AMReX-Codes:development' into development

siddanib committed:
Merge branch 'AMReX-Codes:development' into development

siddanib committed:
Merge branch 'AMReX-Codes:development' into development

Commit:
Importing mpi4py not needed when amrex::MPMD is leveraged.

siddanib committed:
Merge branch 'AMReX-Codes:development' into development

Commit:
First working conversion of mpi4py comm to C MPI_Comm

Commit:
MPMD now depends on mpi4py for python scripts.

Commit:
Input file for lammps_tests/bench/ included

siddanib committed:
Create README.md in tests/test_MPMD

Commit:
A two-way data transfer example is included.

siddanib committed:
Merge branch 'AMReX-Codes:development' into development

siddanib committed:
Merge branch 'AMReX-Codes:development' into development

Commit:
First torch.distributed CPU example on a single node.

siddanib committed:
Update README.md of test_MPMD

siddanib committed:
Second update of README.md in test_MPMD

siddanib committed:
Merge branch 'AMReX-Codes:development' into development

siddanib committed:
Merge branch 'AMReX-Codes:development' into development

Commit:
Including an environment.yml

siddanib committed:
Merge branch 'AMReX-Codes:development' into development

Commit:
Removal of test_MPMD subfolder in tests

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ax3l commented on .gitignore:
```diff
@@ -34,6 +34,24 @@
+## MacOS creates dynamic linker symbol files
+*.dSym
+
+## MacOS crap
```
:sweat_smile: 
```suggestion
## MacOS custom folder attributes
```

siddanib replied:
My bad :sweat:

ax3l replied:
No worries :D

ax3l commented on CMakeLists.txt:
```diff
@@ -115,6 +115,15 @@ foreach(D IN LISTS AMReX_SPACEDIM)
 
+    if(${AMReX_MPI})
+      execute_process(COMMAND
+        ${Python_EXECUTABLE} -c "import mpi4py; print( mpi4py.get_include() )"
```
As a note, this does not work when cross-compiling. I will work on an alternative that does not needs mpi4py to be present and executable at compile time.

- https://github.com/openPMD/openPMD-api/blob/c2e0e83f0380fad0a7996cb6a496d0f66d2d5b6b/src/binding/python/Series.cpp#L38-L52
- https://github.com/openPMD/openPMD-api/blob/c2e0e83f0380fad0a7996cb6a496d0f66d2d5b6b/src/binding/python/Series.cpp#L226-L308

siddanib replied:
Thanks, Axel :+1: ! I included the dependence on mpi4py as my project involves a downstream task that depends on mpi4py. Actually getting just the C++ version of MPI_Comm from https://github.com/AMReX-Codes/amrex/blob/development/Src/Base/AMReX_MPMD.cpp#L97 and directly passing it back to amrex::Initialize worked in my initial attempts.

ax3l replied:
Hi @siddanib,

Yes the approach I linked will work with `mpi4py` :)
It just changes the build process to build without having to provide a run-able `mpi4y` at configuration and compilation time, which is needed when cross-compiling for different architectures.

ax3l commented on .gitignore:
```diff
@@ -34,6 +34,24 @@
 *.out
 *.app
 
+## Compiler debris
```
No big deal here, but for pull requests please try to address one aspect at a time. Changes to this file would be best done in a separate PR :)

ax3l replied:
Moved to #280

siddanib committed:
Update .gitignore

Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>

siddanib committed:
Merge branch 'AMReX-Codes:development' into develop_pr

siddanib committed:
Merge branch 'AMReX-Codes:development' into develop_pr

siddanib committed:
Merge branch 'AMReX-Codes:development' into develop_pr

siddanib committed:
Merge branch 'AMReX-Codes:development' into develop_pr

siddanib committed:
Merge branch 'AMReX-Codes:development' into develop_pr

siddanib committed:
Merge branch 'AMReX-Codes:development' into develop_pr

ax3l committed:
MPI4Py Comm Wrapper

Simplify Cross-Compiles. Using same solution as in openPMD-api to
test and convert MPI Communicators.

ax3l commented:
@siddanib I pushed the update I suggested in review. Can you give it a try if this still works for you? :)

ax3l committed:
Merge remote-tracking branch 'mainline/development' into develop_pr

ax3l commented on src/Base/MPMD.cpp:
```diff
@@ -0,0 +1,171 @@
+              //! -
+              //! https://bitbucket.org/mpi4py/mpi4py/src/3.0.0/demo/wrap-c/helloworld.c
+              //! - installed: include/mpi4py/mpi4py.MPI_api.h
+              // if( import_mpi4py() < 0 ) { here be dragons }
```
```suggestion
              auto m_mpi4py = py::module::import("mpi4py");
              amrex::ignore_unused(m_mpi4py);
```

ax3l committed:
Import `mpi4py`

siddanib committed:
Merge branch 'AMReX-Codes:development' into develop_pr

siddanib committed:
Merge branch 'AMReX-Codes:development' into develop_pr

siddanib committed:
Merge branch 'AMReX-Codes:development' into develop_pr

ax3l commented:
From Slack:

Could you please consider this particular branch for testing the previous approach (https://github.com/siddanib/pyamrex/tree/development).
Could you please use this particular case for testing (https://github.com/siddanib/pyamrex/tree/development/tests/test_MPMD/MultiFab_two_way). `AMREX_HOME` in GNUmakefile needs to be changed.
I installed pyAMReX  as follows:
```
cmake -S . -B build -DAMReX_SPACEDIM="1;2;3" -DpyAMReX_amrex_src=/home/bsiddani/siddanib_github/amrex
cmake --build build -j 8 --target pip_install
```
For the test case:
I first build the `main.cpp` file as follows
```
make -j 10
```
Then I run the MPMD  case as follows
```
mpirun -np 8 ./main3d.gnu.DEBUG.MPI.ex : -np 4 python main.py
```
I will submit two clean examples once we figure out this issue. Please let me know if any additional information is needed. Thanks!

siddanib committed:
Merge branch 'AMReX-Codes:development' into develop_pr

Commit:
test_1 added in tests/test_MPMD/

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Commit:
Removed unused variable bx in tests/test_MPMD/test_1/main.py

siddanib committed:
Merge branch 'AMReX-Codes:development' into develop_pr

Commit:
mpi4py.rc settings in tests were changing thread support level from 3 to 0

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

siddanib commented on tests/test_MPMD/test_1/main.py:
```diff
@@ -8,15 +8,9 @@
 # License: BSD-3-Clause-LBNL
 # Authors: Bhargav Sriram Siddani, Revathi Jambunathan, Edoardo Zoni, Olga Shapoval, David Grote, Axel Huebl
 
-import mpi4py
```
The following lines:
```
import mpi4py
mpi4py.rc.initialize = False
mpi4py.rc.finalize = False
```
were resulting in `MPI initialized with thread support level 0` instead of `MPI initialized with thread support level 3`.

Removing these lines results in `MPI initialized with thread support level 3` and avoids the following error:
```
Abort(3755781) on node 10 (rank 10 in comm 0): Fatal error in internal_Comm_free: Invalid communicator, error stack:
internal_Comm_free(87): MPI_Comm_free(comm=0x7f03e8f576b0) failed
internal_Comm_free(44): Null communicator
```
Need to look into this further.

ax3l replied:
Slack: solved :tada:

siddanib committed:
Merge branch 'AMReX-Codes:development' into develop_pr

siddanib committed:
Merge branch 'AMReX-Codes:development' into develop_pr

siddanib committed:
Merge branch 'AMReX-Codes:development' into develop_pr

siddanib committed:
Merge branch 'AMReX-Codes:development' into develop_pr

siddanib committed:
Merge branch 'AMReX-Codes:development' into develop_pr

ax3l reviewed (approved):
Thank you, LGTM! :)

As a follow-up PR, can you add a `tests/test_MPMD/test_1/README.rst` file that describes the exact steps how to run this test?

ax3l commented:
I have to unbreak the macOS CI separately :)
Just a breaking change in homebrew/pip interaction.
